pkg/condition/condition.go

```diff
@@ -0,0 +1,178 @@
+package condition
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"sync"
+
+	operators "github.com/operator-framework/api/pkg/operators/v2"
+	"github.com/pkg/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kubeWait "k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	patcher "github.com/openshift/windows-machine-config-operator/pkg/patch"
+	"github.com/openshift/windows-machine-config-operator/pkg/retry"
+)
+
+const (
+	upgradeableTrueReason   = "upgradeIsSafe"
+	upgradeableTrueMessage  = "The operator is safe for upgrade"
+	upgradeableFalseReason  = "upgradeIsNotSafe"
+	upgradeableFalseMessage = "The operator is currently processing sub-components. At least one controller is busy."
+	// OperatorConditionName is an environment variable set by OLM identifying the operator's OperatorCondition CR
+	OperatorConditionName = "OPERATOR_CONDITION_NAME"
+)
+
+var (
+	// opCondName is the name of the OLM-managed OperatorCondition resource. Empty if operator is not OLM-managed
+	opCondName string
+	// busyControllers is a representation of a set that holds the names of controllers that are currently reconciling
+	busyControllers map[string]struct{}
+	// opCondBusyControllersLock is a mutex lock to ensure thread-safe access of two shared entities,
+	// the OperatorCondition resource and busyControllers set
+	opCondBusyControllersLock sync.Mutex
+)
+
+// init runs once, initializing global variables
+func init() {
+	opCondName = getOperatorConditionCRName()
+	busyControllers = make(map[string]struct{})
+}
+
+// exported methods
+
+// MarkAsFree removes the given controller from the set of busy controllers, sets the Upgradeable condition to True
+// to allow OLM to perform operator upgrades. No-op if operator is not OLM-managed
+func MarkAsFree(c client.Client, watchNamespace, controllerName string) error {
+	// Check if operator is OLM-managed
+	if opCondName == "" {
+		return nil
+	}
+
+	opCondBusyControllersLock.Lock()
+	defer opCondBusyControllersLock.Unlock()
+
+	// Remove given controller from busy controllers set
+	delete(busyControllers, controllerName)
+
+	opCond, err := get(c, watchNamespace)
+	if err != nil {
+		return err
+	}
+
+	// If at least one other controller is still busy, no-op as Upgradeable should not be set to True yet
+	// Also, if Upgradeable condition is already True, no-op to avoid redundant API calls
+	if len(busyControllers) > 0 || validate(opCond.Spec.Conditions, operators.Upgradeable, meta.ConditionTrue) {
+		return nil
+	}
```

Comment:
This seems weird to combine these. You can check if len(busyControllers) >0 before you do the get.
I'm not sure if this is an effort to solve the issue described [here](https://github.com/openshift/windows-machine-config-operator/pull/618#discussion_r723420711)?

Is this issue actually solved? Don't you always have to validate the opCond?

Reply:
> This seems weird to combine these. You can check if len(busyControllers) >0 before you do the get

Fair point

> Don't you always have to validate the opCond?

In the [`MarkAsBusy` method](https://github.com/openshift/windows-machine-config-operator/pull/618/commits/3d0c2ff167758d578c7601d85b250502d44828a8#diff-4e95dc26dacdca908d7eb173740544f27a74edf6474f9fbbeaa5461f57089fc7R94), yes we always have to validate the operator condition. 
But in the `MarkAsFree` case, we want to be a bit stricter before we can say upgrades are safe. If either of the conditions fails at this point, we cannot set `Upgradeable` to True. Here's all the scenarios:

- if at least one controller is busy (meaning upgradeable _should_ be false) and Upgradeable is currently False
  - no-op
- **important case:** if at least one controller is busy (meaning upgradeable _should_ be false) and Upgradeable is currently True
  - no-op
  - another controller had added itself to `busyControllers` set but hit an error when setting status in OperatorCondition
- if 0 controllers are busy (meaning upgradeable _should_ be true) and Upgradeable is currently True
  - no-op
- if 0 controllers are busy (meaning upgradeable _should_ be true) and Upgradeable is currently False
  - happy path, go ahead and set status in OperatorCondition to True

Reply:
Nice explanation @saifshaikh48. How about expanding the comment to include your explanation in a more succinct manner.